core/src/main/java/org/infinispan/marshall/core/internal/ExternalJBossMarshaller.java

```diff
@@ -0,0 +1,226 @@
+package org.infinispan.marshall.core.internal;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.OutputStream;
+
+import org.infinispan.commons.io.ByteBuffer;
+import org.infinispan.commons.marshall.BufferSizePredictor;
+import org.infinispan.commons.marshall.Externalizer;
+import org.infinispan.commons.marshall.MarshallableTypeHints;
+import org.infinispan.commons.marshall.StreamingMarshaller;
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.marshall.core.JBossMarshaller;
+import org.jboss.marshalling.ByteInput;
+import org.jboss.marshalling.ByteOutput;
+import org.jboss.marshalling.ObjectTable;
+import org.jboss.marshalling.Unmarshaller;
+
+final class ExternalJBossMarshaller implements StreamingMarshaller {
+
+   final MarshallableTypeHints marshallableTypeHints = new MarshallableTypeHints();
+   final JBossMarshaller marshaller;
```

Comment:
As explained last week, we currently have an corner case in our code base that did not allow me to have a version of an external/user marshaller that simply used standard serialization (e.g. org.infinispan.marshall.core.internal.ExternalJavaMarshaller): 

Clustered Query implementation relies on some Lucene classes are Serializable but some of their inner fields are not. We have Externalizers for these inner fields, so in the previous marshaller implementation this work fine. However, when I tried to use a ExternalJavaMarshaller, this logic broke because I didn't have a way to fit in custom serialization logic inside to be able to piggyback to the externalizers. Since that logic is still in place, I made the external jboss marshaller a bit more complex to deal with it, by being able to plug at the `ObjectTable` level and do the corresponding externalizer lookup.

Also, I think we're mixing up some concepts here:

I differentiate between internal and external marshaller. The internal marshaller deals with types we know, either hardcoded (primitives) or via pluggable externalizers. The external marshaller deals with types we don't know. These types might be user defined keys and values, but as you can see above, there might be situations where we deal with 3rd party libraries internally (not only Lucene but you could think of alternative transaction manager implementations). For 3rd party libraries we either rely on standard Java Serialization or we provide Externalizers for those types.

So, I do agree with your point above, but unless clustered query issue is solved, or the functionality is removed, I can't see how I can get around it. @Sanne did say that the clustered query functionality is a little bit flaky, so it'd be very helpful to us to get rid of it.

Any other suggestions?